src/iterator/backend.rs

```diff
@@ -0,0 +1,380 @@
+        let me = Self {
+            ids: Arc::new(RegisteredSignals(Mutex::new(ids))),
+            waker,
+        };
+        for sig in signals {
+            me.add_signal(*sig.borrow())?;
+        }
+        Ok(me)
+    }
+
+    /// Registers another signal to the set watched by this [`Signals`] instance.
+    ///
+    /// # Notes
+    ///
+    /// * This is safe to call concurrently from whatever thread.
+    /// * This is *not* safe to call from within a signal handler.
+    /// * If the signal number was already registered previously, this is a no-op.
+    /// * If this errors, the original set of signals is left intact.
+    /// * This actually registers the signal into the whole group of [`SignalDelivery`]
+    ///   cloned from each other, so any of them might start receiving the signals.
+    ///
+    /// # Panics
+    ///
+    /// * If the given signal is [forbidden][crate::FORBIDDEN].
+    /// * If the signal number is negative or larger than internal limit. The limit should be
+    ///   larger than any supported signal the OS supports.
+    pub fn add_signal(&self, signal: c_int) -> Result<(), Error> {
+        assert!(signal >= 0);
+        assert!(
+            (signal as usize) < MAX_SIGNUM,
+            "Signal number {} too large. If your OS really supports such signal, file a bug",
+            signal,
+        );
+        let mut lock = self.ids.0.lock().unwrap();
+        // Already registered, ignoring
+        if lock[signal as usize].is_some() {
+            return Ok(());
+        }
+
+        let waker = Arc::clone(&self.waker);
+        let action = move || {
+            waker.pending[signal as usize].store(true, Ordering::SeqCst);
+            waker.wake();
+        };
+        let id = unsafe { crate::register(signal, action) }?;
+        lock[signal as usize] = Some(id);
+        Ok(())
+    }
+
+    /// Get a reference to the read end of the self pipe
+    pub fn get_read(&self) -> &R {
+        &self.waker.read
+    }
+
+    /// Drains all data from the internal self-pipe. This method will never block.
+    fn flush(&self) {
+        const SIZE: usize = 1024;
+        let mut buff = [0u8; SIZE];
+
+        unsafe {
+            // Draining the data in the self pipe. We ignore all errors on purpose. This
+            // should not be something like closed file descriptor. It could EAGAIN, but
+            // that's OK in case we say MSG_DONTWAIT. If it's EINTR, then it's OK too,
+            // it'll only create a spurious wakeup.
+            while libc::recv(
+                self.waker.read.as_raw_fd(),
+                buff.as_mut_ptr() as *mut libc::c_void,
+                SIZE,
+                libc::MSG_DONTWAIT,
+            ) > 0
+            {}
+        }
+
+        if self.is_closed() {
+            // Wake any other sleeping ends
+            // (if none wait, it'll only leave garbage inside the pipe, but we'll close it soon
+            // anyway).
+            self.waker.wake();
+        }
```

Comment:
I wonder, if the read end is now accessible through the public API, should this quirk/hack to wake & close all copies of the read end be documented somewhere?

Reply:
I wonder if there is a reason why you didn't use the [`shutdown`](https://doc.rust-lang.org/std/os/unix/net/struct.UnixStream.html#method.shutdown) method for waking and closing all copies. The documentation states that after shutting down the reading side calls to read will return immediately with zero bytes read.

Reply:
If I remember correctly, *this* code makes it close all the copies. Because the shut down wakes *one* of them, so the one that wakes need to close itself and then wake another.

I might be wrong, though, I'm guessing from memory right now.

Reply:
I added some warning to the get_read method that it is no good idea to read from the pipe directly.